crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -1,232 +1,1126 @@
+                    if !children.contains(&system_index) {
+                        children.push(system_index);
+                    }
                 }
+                None => return Err(GraphError::LabelNotFound(label.clone())),
             }
         }
     }
+    Ok(graph)
 }
 
-pub enum ShouldRun {
-    /// No, the system should not run
-    No,
-    /// Yes, the system should run
-    Yes,
-    /// Yes, the system should run and after running, the criteria should be checked again.
-    YesAndLoop,
+/// Generates a topological order for the given graph; panics if the graph cycles,
+/// using given list of system containers to find and print system display names.
+fn topological_order(
+    systems: &[impl SystemContainer],
+    graph: &HashMap<usize, Vec<usize>>,
+) -> Vec<usize> {
+    fn check_if_cycles_and_visit(
+        node: &usize,
+        graph: &HashMap<usize, Vec<usize>>,
+        sorted: &mut Vec<usize>,
+        unvisited: &mut HashSet<usize>,
+        current: &mut HashSet<usize>,
+    ) -> bool {
+        if current.contains(node) {
+            return true;
+        } else if !unvisited.remove(node) {
+            return false;
+        }
+        current.insert(*node);
+        for node in graph.get(node).unwrap() {
+            if check_if_cycles_and_visit(node, &graph, sorted, unvisited, current) {
+                return true;
+            }
+        }
+        sorted.push(*node);
+        current.remove(node);
+        false
+    }
+    let mut sorted = Vec::with_capacity(graph.len());
+    let mut current = HashSet::with_capacity_and_hasher(graph.len(), Default::default());
+    let mut unvisited = HashSet::with_capacity_and_hasher(graph.len(), Default::default());
+    unvisited.extend(graph.keys().cloned());
+    while let Some(node) = unvisited.iter().next().cloned() {
+        if check_if_cycles_and_visit(&node, graph, &mut sorted, &mut unvisited, &mut current) {
+            panic!(
+                "Found a dependency cycle between systems: {:?}.",
+                current
+                    .iter()
+                    .map(|index| systems[*index].display_name())
+                    .collect::<Vec<_>>()
+            )
+        }
+    }
+    sorted
 }
 
-impl<S: System<In = (), Out = ()>> From<S> for SystemStage {
-    fn from(system: S) -> Self {
-        SystemStage::single(system)
+/// Rearranges given vector of system containers to match the order of the given indices.
+fn rearrange_to_order(systems: &mut Vec<impl SystemContainer>, order: &[usize]) {
+    let mut temp = systems.drain(..).map(Some).collect::<Vec<_>>();
+    for index in order {
+        systems.push(temp[*index].take().unwrap());
     }
 }
 
-pub struct RunOnce {
-    ran: bool,
-    system_id: SystemId,
-    resource_access: TypeAccess<TypeId>,
-    archetype_access: TypeAccess<ArchetypeComponent>,
+/// Populates the map of all ascendants and descendants using the given graph.
+fn populate_relations(
```

Comment:
I went ahead and tried to do that, now I remember why I didn't earlier: the exclusive systems don't store resolved dependencies, only the raw labels. So to apply this idea we have to either:
* Make exclusive systems do what parallel systems do, complete with inverted order mapping and retaining the resolved dependencies, which they don't need for anything other than ambiguity detection (which is always one-shot, because everything has to be resolved and sorted again on system insertion/removal anyway).
* Build new graphs from raw labels after exclusive systems have been sorted.
* Leave exclusive systems as-is and do the more efficient thing just for parallel systems.

Reply:
Gotcha. I think (1) is probably my preference to keep code size small. But I'm also down to drop a PERF note in there and we can have this conversation when we decide whether or not to enable this by default. If it is opt-in (or removed entirely), then perf matters less imo.

Reply:
I "accidentally" did just that while refactoring in ec845c5946273c984ff444498daa4750af84ad17. There's more that can be done, but it should do for now.

Reply:
awesome awesome. looking now.